corehq/apps/linked_domain/applications.py

```diff
@@ -101,19 +92,45 @@ def get_linked_apps_for_domain(domain):
     return linked_apps
 
 
-def link_app(linked_app, master_domain, master_id, remote_details=None):
-    DomainLink.link_domains(linked_app.domain, master_domain, remote_details)
-
-    linked_app.family_id = master_id
-    linked_app.doc_type = 'LinkedApplication'
+def _create_linked_app(domain, app_name, upstream_app_id):
+    from corehq.apps.app_manager.models import LinkedApplication
+    linked_app = LinkedApplication(
+        domain=domain,
+        name=app_name,
+        upstream_app_id=upstream_app_id,
+    )
     linked_app.save()
+    return linked_app
+
+
+def create_linked_app(upstream_domain, upstream_id, downstream_domain, app_name=None, remote_details=None):
+    """
+    Links domains if not already linked, creates a LinkedApplication in the downstream domain
+    Handles any additional setup necessary for a newly linked app
+    """
+    DomainLink.link_domains(downstream_domain, upstream_domain, remote_details)
+    if not app_name:
+        original_app = get_app(upstream_domain, upstream_id)
+        app_name = original_app.name
 
+    linked_app = _create_linked_app(downstream_domain, app_name, upstream_id)
+    return handle_special_cases_for_linked_app(linked_app)
+
+
+def _pull_multimedia_if_remote(linked_app):
     if linked_app.master_is_remote:
         try:
             pull_missing_multimedia_for_app(linked_app)
         except RemoteRequestError:
             raise AppLinkError('Error fetching multimedia from remote server. Please try again later.')
 
+
+def handle_special_cases_for_linked_app(linked_app):
```

Comment:
Could these operations be part of `update_linked_app`? A recently-created linked app is going to be updated immediately, isn't it?

Reply:
The specific code smell is the naming/comment - `handle`, `special`, and `misc.` are all little red flags. Yellow flags, perhaps.

Reply:
Yeah it didn't feel good writing this. That sounds right to me, and it is true that everywhere `handle_special_cases_for_linked_app` is used it is followed by `update_linked_app`. That method lives in the app_manager app but I'm tempted to move it into this linked_domain/applications.py file. Any opposition to that?

Reply:
That sounds fine to me.